### ISSUE TITLE
Add support for favorite/pin dags to dashboard.

### ISSUE THREAD

**Issue by tirkarthi** [apache/airflow#45207]
### Description

There is a widget to show recent dags viewed. I was prototyping this by using `useEffect` on dags page on load to add the dag id to a local storage key and display it. The key could store 5 items. I guess it would be useful to allow users to favorite or pin the dags. Simplified approach would be local storage but it won't persist across browsers. There is no model to store this in the server side.

Sample screenshot : 

![image](https://github.com/user-attachments/assets/65d6d876-c82c-4669-aa2a-8c8b05a170be)

### Use case/motivation

Allow users to quickly visit their favorite dags from dashboard.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by tirkarthi** [comment]
Related https://github.com/apache/airflow/issues/21720

**Comment by tirkarthi** [comment]
cc: @pierrejeambrun @bbovenzi

**Comment by bbovenzi** [comment]
Yes, I imagined we would use localStorage for this.

**Comment by pierrejeambrun** [comment]
> Yes, I imagined we would use localStorage for this.

That's definitely the easiest approach but feels like a workaround to avoid having to update the database and do the appropriate backend-work. Basically losing your favorites and therefore having a different dashboard and UI because we swap browser / computer, or used a private navigation.

I don't know. On the other hand doing it in the backend represent a bigger effort.

We can always start small, with the local storage, so the front-end development can progress. And eventually do the backend integration later to persist that if the users request it.

**Comment by tirkarthi** [comment]
Does react provide a way to run a function on every dag page and it's sub pages like dagruns, taskinstances etc ? Something that runs whenever dag/* path matches.

**Comment by pierrejeambrun** [comment]
> Does react provide a way to run a function on every dag page and it's sub pages like dagruns, taskinstances etc ? Something that runs whenever dag/* path matches.

Not that I am aware of. What are you trying to achieve ?

**Comment by tirkarthi** [comment]
Sorry, the comment was slightly off topic and the query was more around use case to build recently viewed dags which was the initial dashboard design. I created this while working on that since I felt favourite dags are more useful than recently viewed dags.

**Comment by pierrejeambrun** [comment]
Basically to get started we need 2 mutations (Create and Delete Favories) and 1 fetch (retrieve favorites).

You can wrap those read-write from the Local Storage into a `rect-query`. This way it's agnostic for our application code if the query is fetching data from the API or from the local storage benefiting from the powerful react-query interface (isLoading, isSuccess, Caching, retries, data sharing and invalidation etc.)

**Comment by bbovenzi** [comment]
We can check url changes via useLocation and useEffect in a wrapper component:

```
const location = useLocation();
useEffect(() => {
   console.log("if contains a dag_id, update view count")
}, [location]);
```

Other DB-supported methods to filter the dashboard could include `owner` or `tags`. The problem is that everyone organizes their dags very differently

**Comment by tirkarthi** [comment]
There seems to be a built in approach from react-query for interface to localstorage and other persistent storage.

https://tanstack.com/query/latest/docs/framework/react/plugins/persistQueryClient

**Comment by rawwar** [comment]
should we limit favourites/pins to DAGs or also allow tasks?

**Comment by bbovenzi** [comment]
> should we limit favourites/pins to DAGs or also allow tasks?

I don't see why not?

**Comment by PNL0** [comment]
Hello,

Could I be assigned to this feature proposal? 

Right now, I have an implementation for this using localStorage, but before I create PR for this I would like to do the appropriate backend work, so the favorite dags persist between browsers.

**Comment by bbovenzi** [comment]
@PNL0  all yours. What do you have in mind for the backend work?

**Comment by PNL0** [comment]
@bbovenzi Thank you. 
Could you assign also my colleague @mariana-marcal-santana? 
We are looking for the best approach for the backend part, but when we have something more concrete, we could share here. 
If you have something in mind, we are open to any suggestion.

**Comment by bbovenzi** [comment]
@mariana-marcal-santana can you comment on the issue? and then I can assign you too.

I am not sure the best way to do this on the backend because we don't have any db concept of user settings which is why I recommended starting with localStorage

**Comment by mariana-marcal-santana** [comment]
Hello @bbovenzi !
I'm @PNL0 's colleague. 
Right now we're still working on the API calls to favorite/unfavorite the DAGs. We'll update with ideas for the backend as soon as possible!

**Comment by PNL0** [comment]
Hello @bbovenzi,

Me and @mariana-marcal-santana have got a solution that we believe that stores the favorite DAGs persistently in the database. 

Instead of using localStorage, the frontend makes requests to favorite the dags and the backend handles this requests with functions that we created.

Could you give some feedback about this approach and check if it matches with what you were describing in https://github.com/apache/airflow/issues/45207#issuecomment-2828581680 ? 

Thanks

**Comment by pierrejeambrun** [comment]
Yes creating a many to many association table for storing `dagid` <=> `userid` seems like the best approach. (to avoid persisting this on the client side)

**Comment by PNL0** [comment]
Hello @pierrejeambrun @bbovenzi 

Couldn't we just update the is_favorite field in the DagModel of each favorite Dag? 
Similar to what is done when you pause a DAG and use the patch_dag API function.

Right now, without any implementation of an association table, I tested the new functionality this way: breeze start-airflow, favorite some DAGs, breeze down, breeze start-airflow again, check if the DAGs are still favorite.  
Since the state of the DAGs didn't change even when I close and reopen the airflow, I assume that the data is persisted in the database, but maybe I'm looking at this in the wrong way and there is other better way to test the functionality.

Also, if the association table is needed, do we also need to create a class that represents the user/user settings?

**Comment by pierrejeambrun** [comment]
> Couldn't we just update the is_favorite field in the DagModel of each favorite Dag?

This field does not exist on the DagModel, I assume you want to create it - The problem with that approach is that all users of the cluster will have the same favorite. It will not be possible to customize favorites 'per user'. I don't think that's what we want.

> I found a definition of User in this file : providers/fab/src/airflow/providers/fab/auth_manager/models/init.py.

Each auth manager has its own User model. We don't need to add one, just the favorite / user relationship, based on the `UserModel` field that represents the identifier. (`BaseModel.get_id` I guess).
